src/graph/immutable_graph.cc

```diff
@@ -976,6 +977,138 @@ SampledSubgraph ImmutableGraph::SampleSubgraph(IdArray seed_arr,
   return subg;
 }
 
+SampledSubgraph ImmutableGraph::LayerSample(IdArray seed_array,
+                                            const float* probability,
+                                            const std::string &neigh_type,
+                                            int n_layers, size_t layer_size) const {
+  unsigned int rand_seed = time(nullptr);
+  auto g_csr = neigh_type == "in" ? GetInCSR() : GetOutCSR();
+  const int64_t* indptr = g_csr->indptr.data();
+  const dgl_id_t* indices = g_csr->indices.data();
+  const dgl_id_t* eids = g_csr->edge_ids.data();
+
+  std::vector<dgl_id_t> nodes;
+  std::vector<int> layer_ids;
+  std::vector<float> probabilities;
+  std::vector<dgl_id_t> edges;
+
+  std::unordered_set<dgl_id_t> candidate_set;
+  std::vector<dgl_id_t> candidates;
+  std::map<dgl_id_t, size_t> n_times;
+
+  size_t n_seeds = seed_array->shape[0];
+  const dgl_id_t* seed_data = static_cast<dgl_id_t*>(seed_array->data);
+  candidate_set.insert(seed_data, seed_data + n_seeds);
+  std::copy(candidate_set.begin(), candidate_set.end(), nodes.begin());
+  layer_ids.insert(layer_ids.end(), nodes.size(), 0);
+  probabilities.insert(probabilities.end(), nodes.size(), 0);
+
+  std::vector<size_t> positions {0, nodes.size()};
+  for (int i = 0; i != n_layers; i++) {
+    candidate_set.clear();
+    for (auto j = positions.end()[-2]; j != positions.back(); ++j) {
+      auto src = nodes[j];
+      candidate_set.insert(indices + indptr[src], indices + indptr[src + 1]);
+    }
+
+    candidates.clear();
+    std::copy(candidate_set.begin(), candidate_set.end(), candidates.begin());
+
+    auto n_candidates = candidates.size();
+    n_times.clear();
+    for (size_t j = 0; j != layer_size; ++j) {
+      auto dst = rand_r(&rand_seed) % n_candidates;
+      if (!n_times.insert(std::make_pair(dst, 1)).second) {
+        ++n_times[dst];
+      }
+    }
+
+    auto n_nodes = static_cast<float>(NumVertices());
+    for (auto const &pair : n_times) {
+      nodes.push_back(pair.first);
+      layer_ids.push_back(i + 1);
+      probabilities.push_back(pair.second / n_nodes);
```

Comment:
why is this sampling probability? also, it's over the number of all nodes in the original graph?